source/common/router/route_config_update_receiver_impl.cc

```diff
@@ -63,7 +63,10 @@ bool RouteConfigUpdateReceiverImpl::onRdsUpdate(const Protobuf::Message& rc,
   for (const auto& vhost : new_route_config->virtual_hosts()) {
     rds_virtual_hosts.emplace(vhost.name(), vhost);
   }
-  rebuildRouteConfigVirtualHosts(rds_virtual_hosts, *vhds_virtual_hosts_, *new_route_config);
+  if (vhds_virtual_hosts_ != nullptr && !vhds_virtual_hosts_->empty()) {
+    // If there are vhosts supplied by VHDS, merge them with RDS vhosts.
+    rebuildRouteConfigVirtualHosts(rds_virtual_hosts, *vhds_virtual_hosts_, *new_route_config);
```

Comment:
Clarifying question/suggestion: if I read [rebuildRouteConfigVirtualHosts](https://github.com/envoyproxy/envoy/blob/456e4497a2fc3d50446d52e61cd5d496f4e1da3f/source/common/router/route_config_update_receiver_impl.cc#L23) correctly, it clears the virtual_hosts (vhosts) from the given `new_route_config`, then adds the vhosts from `rds_virtual_hosts`, which were already in `new_route_config`, and then adds the hosts from `vhds_virtual_hosts_` (both lists are added into `new_route_config.virtual_hosts`).
If the if-condition holds, it means that to get to the same result, only the vhosts in `vhds_virtual_hosts_` need to be added, no?

If that's true, then instead of clearing the vhosts in `new_route_config` + re-adding the `rds_virtual_hosts` vhosts + adding the `vhds_virtual_hosts_` vhosts, this can be replaced by just adding the `vhds_virtual_hosts_` vhosts.

Reply:
rebuildRouteConfigVirtualHosts is called from 2 places -- onRdsUpdate and onVhdsUpdate -- to combine vhost lists coming from both places. The original author(s) went for a simple route of clearing all vhosts and then putting rds and vhds hosts back in, instead of figuring out what exactly needs to be added/updated/removed. I don't think it's worth re-implementing the whole vhost merging logic. The point of this PR is to optimize the common path when vhosts are coming from RDS only, and there is no VHDS.

Reply:
This is just a perf suggestion, as the O(N) re-additions of RDS routes isn't needed, but as you point, this is not the main point of the PR.

Reply:
Yes, I don't want to get into VHDS code paths here, just improving RDS update when there is no VHDS.

Reply:
I think the two copies are just for "merging" rds-pushed vhost and VHDS-pushed vhosts, when either RDS or VHDS are updated. 

Instead of making two copies, it's totally fine to have a range-index indicates how many of the vhosts in the stored RouteConfiguration::vhosts are from RDS/VHDS. 
All the copy around could be move based.

Reply:
Changed to not keep a copy of RDS vhosts at all, if there is no VHDS. This is the largest memory hog.